### ISSUE TITLE
Enable configuring default property values with a lambda expression

### ISSUE THREAD

**Issue by divega** [aspnet/EntityFrameworkCore#11001]
This could be a new overload of `HasDefaultValue` that allows passing a `Expression<Func<T>>` (or even a `Func<TDbContext, T>` if we are ok with drifting from the pattern used in query filters).

This would enable obtaining the value from the actual `DbContext` instance used at SaveChanges time. 

Example:

``` csharp
modelBuilder.Entity<Blog>()
    .Property<string>("TenantId")
    .HasField("_tenantId")
    .HasDefaultValue(() => _tenantId);
```

This should save you from having to write this kind of code:

``` csharp
        public override int SaveChanges()
        {
            ChangeTracker.DetectChanges();

            foreach (var item 
                in ChangeTracker.Entries()
                    .Where(e => e.State == EntityState.Added 
                             && e.Metadata.GetProperties()
                                   .Any(p => p.Name == "TenantId")))
            {
                item.CurrentValues["TenantId"] = _tenantId;
            }

            return base.SaveChanges();
        }
```

**Comment by aradalvand** [comment]
Hey guys, is there any chance this will be considered for v8?
It's a pretty powerful feature and would open the door to lots of possibilities that are currently lacking.

Not sure how complex it is to actually implement, any ideas?

**Comment by TonyValenti** [comment]
I wanted to post here and say that I could really use this feature as well.

In our scenario, we pull data from JSON APIs and cache the responses in a SQLite table. When we do, we cache the raw JSON in a column but also store important properties in their own columns. 

It would be nice to not have to manually map the important properties from the JSON object to the table on add and update and instead be able to let EFCore handle this for us.

**Comment by roji** [comment]
> It would be nice to not have to manually map the important properties from the JSON object to the table on add and update and instead be able to let EFCore handle this for us.

Note that this issue doesn't remove the need to "manually" map your properties; it only allows you to do it via a LINQ expression instead of via SQL.

**Comment by TonyValenti** [comment]
Yes. You are correct. I'd like to do the mapping once when I configure the context and have it automatically re-apply whenever the source data changes.

**Comment by roji** [comment]
So again - you're already able to do that by configuring a computed column as usual, and specifying raw SQL in the computed column definition which fetches data from column (e.g. with the `JSON_VALUE()` function). The only thing this issue tracks is allowing you to express that definition via LINQ, as opposed to via raw SQL. There's nothing here you can't already do.

**Comment by TonyValenti** [comment]
I am hopeful that when this gets implemented, I will be able to express computed properties as provider independent C# expressions as opposed to provider-dependent SQL scripts.

**Comment by lucaslorentz** [comment]
Hey, just want to share that for a few weeks I've working on auto-updated "computed properties" in EF Core: https://github.com/lucaslorentz/computed-expression. Unofficial project!

It doesn't solve the issue reported by @divega, but seems a good fit for @TonyValenti use case
It is still in pre-release, but I would appreciate some early testing and feedback if it is useful and working properly.

---

Update: this is now stable and published as [LLL.AutoCompute.EFCore](https://github.com/lucaslorentz/auto-compute) (v1.6.0, .NET 8/9/10). It maintains persisted computed properties automatically on SaveChanges, with incremental updates, migration backfill, and a web UI to verify and fix inconsistent rows. Unofficial project — feedback welcome.

**Comment by TonyValenti** [comment]
Hi @lucaslorentz - 
This looks like exactly what I'd be wanting.

Does this allow calling custom functions and such in the computed property expression?

**Comment by MGREMY** [comment]
Is there any update on this ? This could be good to have computed value not tied to the database provider behind it

**Comment by roji** [comment]
This issue is in the backlog, and has had only 5 upvotes in its 8-year lifetime, so is rather low-priority.
